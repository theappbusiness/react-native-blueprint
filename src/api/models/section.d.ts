export interface Section {
  title: string;
  body_elements: any[];
}
