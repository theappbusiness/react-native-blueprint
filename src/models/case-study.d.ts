import {Section} from './section';

export interface CaseStudy {
  id: string;
  title: string;
  teaser: string;
  client: string;
  vertical: string;
  hero_image: string;
  sections: Section[];
}
