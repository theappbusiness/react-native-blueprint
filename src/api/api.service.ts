import axios, {AxiosRequestConfig, AxiosResponse} from 'axios';
import Config from 'react-native-config';

const api = axios.create({
  baseURL:
    Config.API_ENDPOINT ||
    'https://raw.githubusercontent.com/theappbusiness/engineering-challenge/main/endpoints/v1',
});

console.log(Config.API_ENDPOINT);

export function get<T>(
  url: string,
  config?: AxiosRequestConfig,
): Promise<AxiosResponse<T>> {
  return api.get<T, AxiosResponse<T>>(url, config);
}

export default api;
