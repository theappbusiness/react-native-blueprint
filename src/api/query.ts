export const QUERY = {
    CASE_STUDIES: '/caseStudies.json',
}