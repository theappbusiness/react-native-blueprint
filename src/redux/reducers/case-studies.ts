import {createSlice} from '@reduxjs/toolkit';
import {CaseStudy} from 'api/models/case-study';

const caseStudySlice = createSlice({
  name: 'case-studies',
  initialState: {
    caseStudies: [] as CaseStudy[],
  },
  reducers: {
    setCaseStudies(state, action) {
      state.caseStudies = action.payload;
    },
  },
});

export const {setCaseStudies} = caseStudySlice.actions;
export default caseStudySlice.reducer;
