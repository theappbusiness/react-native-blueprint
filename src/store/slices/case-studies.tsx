import {CaseStudy} from 'api/models/case-study';
import {StateCreator} from 'zustand';

export interface CaseStudyState {
  caseStudies: CaseStudy[];
  setCaseStudies: (caseStudies: CaseStudy[]) => void;
}

export const createCaseStudySlice: StateCreator<CaseStudyState> = (
  set: any,
) => ({
  caseStudies: [],
  setCaseStudies: (caseStudies: CaseStudy[]) =>
    set(() => ({caseStudies: caseStudies})),
});
