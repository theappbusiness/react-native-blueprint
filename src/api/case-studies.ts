import { get } from './api.service';
import { QUERY } from './query';

export const getCaseStudies = async () => {
  const res = await get<any>(QUERY.CASE_STUDIES);
  return res.data?.case_studies || [];
};