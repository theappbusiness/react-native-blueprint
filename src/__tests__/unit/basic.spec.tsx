import {cleanup} from '@testing-library/react-native';
import {CaseStudy} from 'api/models/case-study';
import axios from 'axios';
import {getCaseStudies} from 'api/case-studies';
import configureStore from 'redux-mock-store';
import {setCaseStudies} from 'redux/reducers/case-studies';

// import {NavigationContainer} from '@react-navigation/native';
// import {render, waitFor} from '@testing-library/react-native';
// import Navigator from 'navigation/RootNavigation';

jest.mock('axios');

const middlewares: any = [];
const mockStore = configureStore(middlewares);
// const addTodo = () => ({type: 'ADD_TODO'});

afterEach(cleanup);

let caseStudy: CaseStudy;

beforeEach(async () => {
  caseStudy = {
    id: '1',
    title: 'Test Case Study',
    sections: [],
  };
});

describe('Basic unit tests', () => {
  // mock test - mocking axios
  it('Shoud fetch the case studies', () => {
    const caseStudies = [caseStudy];

    const mockRes = {
      data: {
        case_studies: caseStudies,
      },
    };

    // axios.get.mockResolvedValue(mockRes);
    axios.get.mockImplementation(() => Promise.resolve(mockRes));

    return getCaseStudies().then(res => expect(res).toEqual(caseStudies));
  });

  it('should dispatch action', () => {
    // Initialize mockstore with empty state
    const initialState = {
      caseStudies: [] as CaseStudy[],
    };

    const store = mockStore(initialState);

    // Dispatch the action
    store.dispatch(setCaseStudies([caseStudy]));

    // Test if your store dispatched the expected actions
    const actions = store.getActions();
    const expectedPayload = {
      payload: [caseStudy],
      type: 'case-studies/setCaseStudies',
    };
    expect(actions).toMatchSnapshot();
    expect(actions).toEqual([expectedPayload]);
  });
});
