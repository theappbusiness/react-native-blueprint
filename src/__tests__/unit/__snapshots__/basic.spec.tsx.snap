// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic unit tests should dispatch action 1`] = `
[
  {
    "payload": [
      {
        "id": "1",
        "sections": [],
        "title": "Test Case Study",
      },
    ],
    "type": "case-studies/setCaseStudies",
  },
]
`;
